[1.4] Accessory Slot Creation Rework

Solxanich opened the pull request:
### What is the new feature?
Credit to ModernMak for initially working on cleaning up the accessory code, which gave me the idea to actually make extra accessory slots easier for modders.
Well, currently adding accessory slots is not a clean affair. Lot's of cloning, netmessaging, drawing, and that just gets you the functional slot. 
It ends up being a lot of boilerplate code, doing the same stuff in multiple mods.
Ignoring that, not all mods cover all functionality that a slot has, vanilla does some fun stuff thats easy to miss, such as post load fixes.
The ability to retrieve your items when you unload the mod is lost.
Finally, we have the issue of UI cluttering from such slots. 

This PR aims to fix all of that by making a ModAccessorySlot ModType. An example of the player facing UI behavior is as follows (1 custom location slot, and two additional standard drawn slots): 


https://user-images.githubusercontent.com/59670736/133655350-f178d1a4-f9ff-478c-9899-71d6ff909e7c.mp4


### Example Usage:
See ExampleMod ExampleAccessorySlots for usage such as:
![image](https://user-images.githubusercontent.com/59670736/133657353-38f0f0cd-2b0b-467b-bee4-31493da73d6c.png)


 ### Concerns:
The interdependency of AccessorySlotLoader.list.Count & ModAccessorySlotPlayer array.Length while dealing with multiple different player characters with varying unloaded slots may result in errors in Multiplayer. 
Non-trivial to test locally, so this particular scenario is untested and possible problem.
There are possible solutions, however, would prefer to avoid implementing either unless this is an actual problem and can see what in particular is the breaking point to select the appropriate fix path around.

ThomasThePencil committed:
fixed DamageClass throwing an exception before the game can even start

if this gets undone heads are gonna fucking roll

Solxanich committed:
Merge pull request #10 from ThomasSpeedrunner/1.4_damageclassfix

fixed DamageClass throwing an exception before the game can even start

Commit:
Adds static fields and values to make accessories and inventory dynamic.
Replaces many hardcoded numbers with equivalent fields
Can't open player screen properly. IsBugged

Commit:
Incomplete; realizing got carried away.
Gonna shelf this branch.

Branch has a mostly implemented dynamic accesory slots counting; meant to make adding accessory slots easier, but is getting out of hand, so gonna refocus in on original task.

Commit:
Revert "Merge pull request #10 from ThomasSpeedrunner/1.4_damageclassfix"

This reverts commit 808d7ebe68bdd01977140c0eb8e1f09518bc00f4, reversing
changes made to 1ab1050f1834076590a296017dc45be0e240c687.

Solxanich committed:
Merge pull request #11 from tModLoader/1.4

1.4

Commit:
De-shelved. I think there is meaning here in making modded accessory slots actually clean.
Fixed a bunch of items; notably made vanilla saving be only vanilla slots again, as it should be.
Branch as is, is just a structural rework of vanilla naming that doesn't do anything.
PR now, or later?

Commit:
Finished testing that all vanilla movements, values, saving, loading work fine without modded slot being non-zero. Good start.

Commit:
2 Issues to resolve for next commit:
1) in cases 9,11 in RightClickSpecial the swap is going wrong and upshifting by totalEquipSlots - vanillaEquipSlots. Funky
2) The armor vanity slots are not being saved in player.cs. Accesory vanities are being saved correctly, however.

Commit:
Calling this branch: Accessory Slot Creation Rework; Part 1: The Vanilla Twist.
Mostly implemented.
Remaining issues of unable to leftclick/shiftleftclick to place armor into vanity slot. Mostly because didn't deal with it yet.
Multiplayer is working wellish; draw issues with multiplayer exclusive UI draws such as team colours.

Chicken-Bones reviewed (commented):
tl;dr: skip to the last paragraph.

Thanks for making this PR as a partially implemented proposal. This lets us evaluate the maintenance impacts of your suggested approach, and discuss actual code. Let me start by saying that I'm definitely in favour of tML providing an easy mechanism to add extra accessory slots. The picture you've provided is exactly what we should hope to achieve with this PR. Please see the individual code review comments below.

While you've noted that save/load and hooks for 'customisation' of slots are not yet implemented, we need to discuss these to settle on a design for this feature. The current hook design does allow mods to create their own slots, but they must handle rendering, storage, networking and interactions, replicating a lot of vanilla functionality. 

Normally, if a mod adds an accessory slot, and the mod is unloaded, the accessory in it is inaccessible until the mod is restored. With tML handling this feature, we can save accessories centrally, while retaining which mod owns their 'slot', and if the mod is disabled (or the slot removed) then we can use the vanilla 'no effect' rendering to allow players to remove their accessories from the missing slots and then have the slots disappear.

I also expect mods to want to restrict what can be placed in their slots, or violate certain normal rules (like allowing duplicate accessory use, or making a slot which applies effects twice).

With this in mind, the number of accessory slots cannot simply be a public variable on player. Mods must declare (possibly via a `ModEquipSlot` class) their slots, for tML to internally sort, render, store and network. 

The approach here, of storing mod accessory slots in the same arrays as vanilla, requires changing many constants, and hooking many IO and render locations, incurring significant maintenance cost. Additionally, mods are still going to want to create 'special' slots outside these arrays, and I fear that relying too heavily on vanilla code paths may leave them out on interoperability.

Please see my mod https://github.com/Chicken-Bones/UtilitySlots (which I updated just for this review), for an example of exactly what building a current equipment slot entails. UtilitySlots is a little bit special in that it adds its own 'page', but it should illustrate what's currently difficult and missing. In particular, note that an `On` hook is required for `Player.UpdateVisibleAccessories` and that there's currently no way interact with the fast equip feature (`ArmorSwap`) without an IL hook.

I think the best approach will actually be for tML to implement the extra slots using available hooks, and the builtin `ModLoader` mod. This will mean minimal patches, and that tML is using the same API to create the shared pool of 'plain extra accessory slots' as mods can use to make more specialised content.

Solxanich commented:
Wow, thank you for the expedited review! I wasn't expecting a review for a bit as I had labeled this draft and was still cleaning up things here and there, as well as checking for items missing in description. Glad to hear their is an useful outcome here, although design is up for discussion.

I guess to address in order, I'll start by adding the info on what I was thinking for part 2 that I clearly missed adding to original desc. Part 2 I was thinking would:
1) Swap the integer to a proper tagcompound list with data:
A) index of assigned slot -> so that a mod can override and custom draw of the slot, as well as modify slot using existing hooks
B) name of mod -> so can check against and not display when unloaded
C) a pointer to the mod's player class to obtain data such as ShouldDisplay for character and Custom Conditions on placement -> but I suspect I'd have to implement differently this to be cleaner than initial idea
2) Add a single, unified load method and single save method to handle the additional slots data, instead of all overall the place like vanilla is
3) check code to ensure functionality of two above is fully implemented.

For particular code, agree with all flagged stuff; missed how we handle vanilla disabling, so I'll clean that up if continuing on this design path.

For the comment, by paragraph:
1) Agree - save and load is important. My thoughts are tML should self manage saving and loading of additional slots in terms of how many, and what items were there. Custom conditions, custom draws should be handled by the mod imo
2) See part 2 notes (again, sorry, should've waited before creating PR so that I could spend time ensuring all notes are on it instead of sleeping on the draft)
3) Been basing initial thinking off of WingSlot. So yes, see part 2 notes, 1c. Hadn't thought about intention break normal rules, so good catch on that.
4) I'll need to think on this when it comes to custom slots. In terms of maintenance, agree lots of number swaps to variables, although not sure maintenance created here. The assumption I'm making is most of vanilla code is basically frozen, never going to change, at this point so I'm not sure maintenance in terms of updating with Vanilla is something I'd consider personally. 
5) I'll take a look at your mod. Thank you for updating it; good to have a relevant comparison
6) Yeah, alternatives definitely exist. I'll take a look at going the modloader route with its own management of slots instead; I imagine it will still accomplish general same idea, although with a bit longer blocks of custom tML code instead of many small vanilla patches.

Tl;dr I jumped the gun on posting PR without verifying all details presented.
Also think we should review our approach to vanilla patches given vanilla code is hard-frozen now (exempting steam workshop and related minor changes)

Solxanich commented on patches/tModLoader/Terraria/Main.cs.patch:
```diff
@@ -2423,6 +2432,130 @@
++						|| (num40 == player[myPlayer].masterModeSlot && !flag5))
+ 						continue;
++					if (num40 == player[myPlayer].vanillaEquipSlots) // begin a new column for modded accessory slots, with functionals being unique
++						modRowCorrect = num39 + 1;
```
I actually disagree with being in a separate column, despite implementing as such, as it leads to more unnecessary patches relative to the intended feature (namely modifying draws for defense, multiplayer team selections, etc). 
After seeing your UtilitySlots mod, I think it best to implement as a new page, and maybe columns subsequent from there. 
Further, we shouldn't just continually stack downwards in vanilla fashion, as that will inevitably lead to slots rendering offscreen, and thus vanilla's Purple Screen of Render Error. (For example, modSlots >= 11 will purple screen as I don't do more than one column...)

Chicken-Bones replied:
I'm biased against a separate page, but looking at the vanilla UI there's really no space there. Even with another page you'll only get 7 more slots though. I think the best idea is to do what you've done with the stacking up of new accessory slots in another column, and just moving the MP icons and armor around

Solxanich commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -48,15 +48,42 @@
++		public int moddedInventorySlots = 0;
++		public int totalInventorySlots => moddedInventorySlots + vanillaInventorySlots;
++		//public bool extraAccessory; -> renamed to demonHeartUsed to avoid confusion
++		public bool demonHeartUsed;
```
Agreed - this goes back to the same point Can"InsertVanilla"AccessoryBeShown

Solxanich commented on patches/tModLoader/Terraria/ModLoader/ItemLoader.cs:
```diff
@@ -1244,26 +1244,31 @@ private class HookList
 		/// Returns the wing item that the player is functionally using. If player.wingsLogic has been modified, so no equipped wing can be found to match what the player is using, this creates a new Item object to return.
 		/// </summary>
 		public static Item GetWing(Player player) {
```
I'll double check it does nothing otherwise in the player.cs, but yeah, when looking at it, it's very redundant and overly-complicated. 
I'll look to remove it independent of design path.

Solxanich replied:
I haven't removed this method as it is used in horizontal and vertical wing speed fetches in ItemLoader. For now, I don't touch ItemLoader.cs, so I don't think I'll tackle that here.

Solxanich replied:
Resolved by storing "equippedWings" on the player at same time as setting winglogic.

Solxanich commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -48,15 +48,42 @@
++		public int vanillaHotbarSlots = 40;
++		public int vanillaCoinAmmoSlots = 8;
++		public int vanillaInventorySlots => vanillaHotbarSlots + vanillaCoinAmmoSlots + vanillaContainerSlots;
++		public int moddedInventorySlots = 0;
```
I'll remove this. Agreed that it just makes things more complicated when too many unnecessary things included.

Solxanich commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -48,15 +48,42 @@
+ 		public const int SupportedMiscSlotCount = 5;
+ 		public Item[] armor = new Item[20];
+-		public Item[] dye = new Item[10];
++		public Item[] dye = new Item[20];
```
This is a Homer Simpson "Doh!".
Wonder when I changed that; I'll fix that to not change

Commit:
De-rename several items:
1) demonheartUsed -> changed to masked
2) changed vanilla code to be commented out
3) changed CanAccessory to restore vanilla code, and then commented out for forwarders.

Commit:
Inbetween state push to keep content updating.
Adds abstract class ModAccessorySlot, as a tentative subclass of ModType. Moslty hardcode; doesn't do anything right now.
Also some minor cleanup in Main.cs.patch because of this.

Commit:
Pushing a stable version.
Doesn't do anything;
Need to add to ModAccessorySlot class:
1) an overridable method for custom conditions.
Need to add to Player.cs:
1) Full support for custom conditions
Need to add to Main.cs:
1) Complete the ModAccessorySlot Draw call.
Need to create an instance of ModAccessorySlot in exampleMod to test

Chicken-Bones commented:
> Tl;dr I jumped the gun on posting PR without verifying all details presented.
As mentioned before, this was a good state to create a draft in and get feedback.

> The assumption I'm making is most of vanilla code is basically frozen, never going to change, at this point so I'm not sure maintenance in terms of updating with Vanilla is something I'd consider personally.
> Also think we should review our approach to vanilla patches given vanilla code is hard-frozen now (exempting steam workshop and related minor changes)
This is not currently the stance of the tML dev team. Even if we could guarantee that the vanilla code was going to be hard frozen, improvements to ILSpy and Roslyn based reformatters are still on the table. Even if we were more open to sweeping changes, we prefer keeping the original vanilla code for comparison, and creating new code paths which better model the goals of tML. That and vanilla code is often a fragmented mess and difficult to reason about.

Commit:
Need to pause for the day. Hit a roadblock
Current state: can launch and run as vanilla basically.
Current issues: functional except for the fact that player[myPlayer] in main does not account for post mod loading player data changes. So drawing doesn't work with mAccSlot and related loop.
Need to think about.

Commit:
In progress commit to bring branch up to date:
1) Began de-converting vanilla magic numbers
2) Moved equipment related hooks from ItemLoader to EquipLoader for consistency
3) Began to construct EquipLoader to handle ModAccessorySlots; slowly recreating vanilla methods as needed. All points of interest should be flagged by errors in current version.
4) Added ExampleMod exampleAccessorySlot file. Still have to work further before can flesh out.

Solxanich commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -1119,8 +1160,57 @@
+ 		}
+ 
+ 		public bool IsAValidEquipmentSlotForIteration(int slot) {
++			/*
```
I'll need to fix this

Commit:
Small update to bring to a return to later state.

Commit:
Further cleaned up former vanilla changes.
Getting closer; remaining is to work on ModAccessorySlot, EquipLoader and related to work on draw functionality and multiplayer support.

Commit:
Resolved merge conflicts

Solxanich committed:
Merge pull request #15 from Solxanich/tModLoader-1.4

T mod loader 1.4

Commit:
Added Draw functionality.
Remaining: Networking/MP and GamePad support.

Solxanich committed:
Merge pull request #16 from Solxanich/tModLoader-1.4

Added Draw functionality.

Commit:
'Basic' accessory slot functionality in singleplayer has been established.

2 known bugs of crashing on first load with change/ crashing on unload because I'm not being smart enough on handling nulls in Initialize().

Need to add Multiplayer netcode, and test thoroughly all virtuals in ModAccessorySlot and all functionality vectors.

Commit:
Further implemented multiplayer functionality; still need to do netcode, but it appears all functionality is there.

Commit:
Cleaned up a large amount of remnants from previous designs.
No new changes in this commit - all cleanup.

Commit:
Added: Basic functionality for scrollbars
Changed: Moved ModAccessorySlot to be its own subset of ModType to improve performance.

Commit:
Added rudimentary scrolling implementation

Commit:
Added multiplayer netcode, and cleaned up the UI a bit.
This branch is now presentable for review, sort of. Still concerned about how much code there is and refactoring it.

Commit:
Added multiplayer netcode, and cleaned up the UI a bit.
This branch is now presentable for review, sort of. Still concerned about how much code there is and refactoring it.

Commit:
Amended a scrollfix for when going from 6 to 5 slots.

Commit:
Changed UI to have scrolling vanilla accessories. Looks cleaner.

Commit:
Minor cleanup item.

Commit:
Move scrollbar to right, and make slots per column calc dynamic.

Commit:
Final Cleanup Part 1 of 2

Commit:
Cleanup part 2 of 2: Cleanup Player.cs

Commit:
Uses PR #1312 now

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/DefaultPlayer.cs:
```diff
@@ -0,0 +1,183 @@
+
+namespace Terraria.ModLoader.Default
+{
+	public class DefaultPlayer : ModPlayer
```
`ModAccessorySlotPlayer` please

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/EquipLoader.cs:
```diff
@@ -80,12 +87,12 @@ public static class EquipLoader
 			LoaderUtils.ResetStaticMembers(typeof(ArmorIDs), true);
 			WingStatsInitializer.Load();
-			
+
```
accidental removal of spaces?

Solxanich replied:
Hmm. This is a swap from somebody hitting tab 3+ times to just a new line.
Accidental, probably, but I think is better to use just a new line and not new line + multiple tabs.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/EquipLoader.cs:
```diff
@@ -298,5 +305,487 @@ public static class EquipLoader
+			}
+
+			return mAccSlot;
+		}
```
I think a different class would be better for these methods. The role of `EquipLoader` is just mapping item ids to equip types and ids. Mod accessory slot loading/hooks should go in a new class.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,83 @@
+	/// </summary>
+	public abstract class ModAccessorySlot : ModType
+	{
+		internal virtual bool skipRegister => false;
```
needs a better name? Something about not being part of the vanilla slot list, and being manually rendered/positioned by a mod?

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,83 @@
+		/// Called when the ModAccessorySlot is registered. Registers to parent list.
+		/// Requests for a modded Slot, and stores assigned Slot. DefaultPlayer will auto-manage the items themselves.
+		/// </summary>
+		internal protected void Initialize() {
```
`private`?
`internal` allows accesses from tML outside the `ModAccessorySlot` class
`protected` allows accesses from subclasses

Just merge this into `Register`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,83 @@
+		/// <para><paramref name="position"/> :: is the position of where the ItemSlot will be drawn </para>
+		/// <para><paramref name="context"/> :: 12 => dye; 11 => vanity; 10 => functional </para>
+		/// </summary>
+		public virtual bool DrawModded(Item[] inv, int context, int slot, Vector2 position) {
```
normally I'd say call this `CustomDraw` or something, but actually, just move the `ItemSlot.Draw` call from `DrawRedirect` in here, and then if mods override this, they'll be specifying custom behaviour. Just like inheritance was meant to be :)

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,83 @@
+		/// <summary>
+		/// Override to set conditions on when the slot is available, Example: the demonHeart is consumed and in Expert mode in Vanilla.
+		/// </summary>
+		public virtual bool CanUseSlot() {
```
`IsSlotAvailable`? perhaps as a property? summary should mention that this controls _visibility_ and that if the condition is false but an item was previously inserted, the slot will disappear when the item is removed.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,83 @@
+		internal static int scrollbarSlotPosition = 0;
+
+		// List of registered identifiers for modded accessory slots. Used in DefaultPlayer.
+		internal static List<string> moddedAccSlots = new List<string>();
```
put this in ~`EquipLoader`~ `ModAccessorySlotLoader`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,83 @@
+		/// Receives data:
+		/// <para><paramref name="checkItem"/> :: the item that is attempting to enter the slot </para>
+		/// </summary>
+		public virtual bool LimitWhatCanGoInSlot(Item checkItem) {
```
`CanAcceptItem`?

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -1491,13 +1482,13 @@
 +
 +				ItemLoader.UpdateEquip(armor[k], this);
-+			}
+ 			}
```
accidental indentation change? or fix?

Solxanich replied:
Fix, I believe. Instead of us adding the closing '}', we now use the existing vanilla '}'.
I'll take another look for any that aren't a fix.

Chicken-Bones commented on patches/tModLoader/Terraria/UI/ItemSlot.cs.patch:
```diff
@@ -36,6 +36,29 @@
+ 			bool flag = Main.mouseLeftRelease && Main.mouseLeft;
+ 			if (flag) {
+-				if (OverrideLeftClick(inv, context, slot))
++				if (OverrideLeftClick(inv, Math.Abs(context), slot))
```
needs a comment

Chicken-Bones reviewed (changes requested):
Looking good so far. I'll need to take another pass through the actual player and equip slot logic changes. 

Also: how does ModAccessorySlot work with a custom draw position/condition (eg for UtilitySlots or a mod that wants a special slot somewhere). Can we get an example?

Commit:
Update with tested version of UI scrollbar fix.

Commit:
Address Merge  Conflicts

Commit:
Part 1 of 2: Address PR comments on organization.
Noted that some of the player Logic isn't quite right in terms of swapping, checking accessory conflicts, etc. Gonna need to take another look, suspect it is an ordering issue.

Solxanich commented on ExampleMod/Common/Players/ExampleAccessorySlot.cs:
```diff
@@ -0,0 +1,35 @@
+	{
+		// If the class is empty, everything will default to a basic vanilla slot.
+	}
+
```
Add CustomDrawAccessorySlot : ModAccessorySlot with overriding XLoc, YLoc and DrawModded

Solxanich commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,69 @@
+		}
+
+		/// <summary>
+		/// This function allows for custom textures and colours to be drawn for the accessory slot. Called for Dyes, Vanity, and Functionals.
```
Double check the comments to ensure are accurate

Solxanich commented on patches/tModLoader/Terraria/ModLoader/AccessorySlotLoader.cs:
```diff
@@ -0,0 +1,510 @@
+
+			Player player = Main.LocalPlayer;
+			player.mouseInterface = true;
+			player.mouseInterface = true;
```
Uhh, double take?

Solxanich committed:
Finish drafting up changes to ModAccessorySlot Redux

I believe the only thing outstanding now is  possibly documentation being out of date here and there, and an examplemod improvement.

Solxanich committed:
Resolve merge commits

Solxanich committed:
Merge pull request #22 from Solxanich/tModLoader-1.4

T mod loader 1.4

Solxanich committed:
Cleanup absurd Patch by using GoTo

Solxanich committed:
Merge branch '1.4_Extra_acc_fix_redux' of https://github.com/Solxanich/tModLoader into 1.4_Extra_acc_fix_redux

Solxanich committed:
Address-merge-conflicts

Solxanich committed:
Merge pull request #25 from Solxanich/tModLoader-1.4

T mod loader 1.4

Solxanich committed:
Revert "Merge pull request #25 from Solxanich/tModLoader-1.4"

This reverts commit 2961a50f91d4ba686a31fa24b4754fcc68b61a63, reversing
changes made to d0d0cefd1e42177e55ed18c163da5870953336ce.

Solxanich committed:
Revert "Revert "Merge pull request #25 from Solxanich/tModLoader-1.4""

This reverts commit 9f4a328cd099b8738e88a9edcc7319e4627fe932.

Solxanich committed:
Merge branch '1.4' of https://github.com/tModLoader/tModLoader into tModLoader-1.4

Solxanich committed:
Merge pull request #30 from Solxanich/tModLoader-1.4

T mod loader 1.4

Solxanich committed:
Update to 1.4.2.1

Solxanich committed:
Merge branch '1.4_Extra_acc_fix_redux' of https://github.com/Solxanich/tModLoader into 1.4_Extra_acc_fix_redux

Solxanich committed:
Address patch errors.

Still need to look at https://github.com/tModLoader/tModLoader/pull/1370/files

Solxanich committed:
Address Merge conflicts part 1

Solxanich committed:
Update Player.cs.patch

Solxanich committed:
Merge pull request #35 from Solxanich/tModLoader-1.4

T mod loader 1.4

Solxanich committed:
Fix bad merge result

Solxanich committed:
Initial Merge Results

Solxanich committed:
Merge branch 'tModLoader-1.4' into 1.4_Extra_acc_fix_redux

Solxanich committed:
Manually Fix Patch Offsets being broken

Solxanich committed:
Fix Build Errors and Fresh Diff

Solxanich committed:
Cleanup a lot of code to meet my modern standards better

Still a lot TODO.
1) Fix bugs in Armor Swap for right click in inventory.
2) Display accessories in rendering on player selection screen
3) Multiplayer changes for unloaded slots, don't think the current unloaded slots will work in multiplayer too well
4) Cleanup in the Functionality code spread about AccessorySlotLoader and ItemSlot

Solxanich committed:
Draw code refactoring

Solxanich committed:
Add custom texture functionality

Remove ArmorSwap so that can rewrite it

Solxanich committed:
Fix Dye rendering and ArmorSwap

Just dealing with multiplayer & unloaded slots to go

Solxanich committed:
Finish point-testing multiplayer and patch

Solxanich committed:
Finalization of ExampleMod examples and testing

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,86 @@
+			type = Loader.Register(this);
+		}
+
+		private bool MySlotContainsAnItem() {
```
Suggestion, invert this and call it `IsEmpty`
Can use the suggested properties above.
Should probably also check dye slot.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,86 @@
+		internal int index;
+		public AccessorySlotLoader Loader => LoaderManager.Get<AccessorySlotLoader>();
+
+		// Properties to preset a location for the accessory slot
```
There should probably be get+set properties for `Item`, `VanityItem`, `Dye` and `ShowVisuals`

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -1672,7 +1685,7 @@
  
  		private void ApplyEquipVanity(int itemSlot, Item currentItem) {
-+			ApplyEquipVanity(currentItem); //remove itemSlot parameter and make public so mods can call from their own accessory slots
++			ApplyEquipVanity(currentItem);
```
accidental comment revert?

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -1644,11 +1656,12 @@
  
-+			/* wing loop is merged into ApplyEquipFunctional
++			/*
++			// wing loop is merged into ApplyEquipFunctional
```
accidental comment formatting change?

Chicken-Bones commented on patches/tModLoader/Terraria/UI/ItemSlot.cs.patch:
```diff
@@ -260,40 +412,145 @@
+ 				player.armor[num + 2] = item.Clone();
+ 			}
+ 			else if (item.accessory) {
++				//TML: Rewrote ArmorSwap for accessories under the PR #1299 so it was actually readable. No vanilla functionality lost in transition
```
preferred approach to a whole method rewrite is to make another method and call it while commenting out the vanilla impl

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/UnloadedAccessorySlot.cs:
```diff
@@ -0,0 +1,25 @@
+			return false;
+		}
+
+		// Only display unloaded slots if in singleplayer. Don't need people carrying items in to multiplayer they shouldn't have :(
```
unsure about this, may cause confusion. The item can switch to 'unloaded item' independently of the slot. I think if a server removes a mod, or a player joins a server which doesn't have the mod adding the slot, they should still be able to get the item and choose how/if to use it with the other slots they have available here.

Solxanich replied:
Gonna make a judgement call - if a player has an op accessory going in, odds are they have other op items elsewhere on the player.
This kind of check would thus be redundant.
Will default back to just if there is an item

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Default/ModAccessorySlotPlayer.cs:
```diff
@@ -0,0 +1,206 @@
+			}
+		}
+
+		// The following netcode is adapted from Chicken-Bone's UtilitySlots:
```
*ChickenBones'

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/AccessorySlotLoader.cs:
```diff
@@ -0,0 +1,536 @@
+
+			if (modded) {
+				ModAccessorySlot mAccSlot = list[slot];
+				customLoc = mAccSlot.XLoc != 0 || mAccSlot.YLoc != 0;
```
I don't suppose there's any actual desire for custom location (0, 0)? Might be better to return -1 for 'no custom location'? Not a big deal either way.

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -1608,15 +1614,21 @@
- 			}
++			}
++
++			LoaderManager.Get<AccessorySlotLoader>().VanillaUpdateEquipsMirror(this);
```
I'd prefer if all equip related `AccessorySlotLoader` functionality was implemented via `ModAccessorySlotPlayer `. That way we can ensure feature parity with mod accessory slots, and the `ModPlayer` api in terms of having equipment effects/visuals applied from custom sources.

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -935,6 +927,20 @@
+ 					UpdateItemDye(i < 10, hideVisibleAccessory[num], armor[i], dye[num]);
+ 				}
+ 			}
++			LoaderManager.Get<AccessorySlotLoader>().VanillaUpdateDyesMirror(this);
```
See below comment about using a `ModPlayer`

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -3063,6 +3091,18 @@
+-			if (!dead)
++			if (!dead) {
+ 				UpdateVisibleAccessories();
++				LoaderManager.Get<AccessorySlotLoader>().VanillaUpdateVisibleAccessoriesMirror(this);
```
See above comment about using a `ModPlayer`

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -4117,6 +4167,16 @@
+ 			}
+ 
++			if (LoaderManager.Get<AccessorySlotLoader>().VanillaPreferredGolfBallMirror(ref projType))
++				return;
```
See above comment about using a `ModPlayer`. This one probably needs a new hook

Chicken-Bones commented on patches/tModLoader/Terraria/Player.cs.patch:
```diff
@@ -4814,10 +4874,22 @@
  
  				newPlayer.lavaTime = newPlayer.lavaMax;
  			}
++			LoaderManager.Get<AccessorySlotLoader>().VanillaLastMinuteFixesMirror(newPlayer);
```
See above comment about using a `ModPlayer`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/AccessorySlotLoader.cs:
```diff
@@ -0,0 +1,536 @@
+				}
+				
+				if (type == 158 || type == 396 || type == 1250 || type == 1251 || type == 1252)
+					newPlayer.noFallDmg = true;
```
Rather than duplicating the vanilla code here, it would be ideal if the approach used for the rest of the equipment effects was followed. By which I mean:

Split the vanilla `LastMinuteFixes` method into one that takes an `Item` parameter, and the vanilla accessory slot iteration just calls that one. That way mods can call it too.

Then have a `LastMinuteFixes` hook on `ModPlayer` which you use to apply your custom accessory slots.

Chicken-Bones reviewed (changes requested):
Much improved! My notes are mostly around naming, and hooks in `Player`

direwolf420 commented:
```cs
bool flag12 = false;
for (int num23 = 3; num23 < 10; num23++) {
    if (armor[num23].stack > 0 && armor[num23].wingSlot > -1)
        flag12 = true;
}
```

And `!flag12` instead of `wingsLogic == 0`.

Should be present where the fall damage is applied/calculated in Player.cs. More concrete:
Application: `Hurt(PlayerDeathReason.ByOther(0), num25, 0);`, the if conditional leading up to it should be adjusted.
Missing Code: `if (stoned) {`, above it, is where the above vanilla code is located.

direwolf420 commented:
I know this was 6 years ago, but the above mentioned code is what you removed in this commit @Chicken-Bones: [commit](https://github.com/tModLoader/tModLoader/commit/60fbe9f708acc8cc81a34b13eb4ddcd89a654b63), do you remember why it had to be done like that? Was the bug with mounts not present back then?

Chicken-Bones commented:
> I know this was 5+ years ago, but the above mentioned code is what you removed in this commit @Chicken-Bones: [commit](https://github.com/tModLoader/tModLoader/commit/60fbe9f708acc8cc81a34b13eb4ddcd89a654b63), do you remember why it had to be done like that? Was the bug with mounts not present back then?

A poor assumption back then. I did not test with mounts. My intention was to replace the iteration loop with something that could check if a wing effect had been applied this frame (from any source). We still need a proper fix for this. As I mentioned in my review to solxan, effects should come from _items_ not from accessory slots specifically. Mods should be able to feed items into the `UpdateX` methods to produce effects each frame, regardless of if the item actually exists, or is in an accessory slot. 

One solution may be an additional flag in `Player` for 'hasWings' which is set when applying equip effects for a wing even if mounted.

Solxanich committed:
Address all the easy to do feedback items

Solxanich commented:
> I know this was 5+ years ago, but the above mentioned code is what you removed in this commit @Chicken-Bones: [commit](https://github.com/tModLoader/tModLoader/commit/60fbe9f708acc8cc81a34b13eb4ddcd89a654b63), do you remember why it had to be done like that? Was the bug with mounts not present back then?

I've updated the logic to now check the "Item equippedWings" field that this PR adds. Since this field would be accurate independent of mounts, it is much safer to work off of, and can be set by whatever, wherever.

Solxanich committed:
Further Cleanup of duplicate code

 Getting closer to a ModPlayer implementation of the 4? functionality methods

Solxanich committed:
Merge pull request #58 from tModLoader/1.4

1.4 catchup

Solxanich committed:
A rounds of testing & fixes

Solxanich committed:
ModPlayer Migration

Solxanich committed:
De-netcode Unloaded Slot

Solxanich committed:
Address Some feedback from discord for features

1. How can I make my slots enabled but not visible (they render with custom location, with a folding 'visibility toggle')
2. How to control swap order? (Wings should go in the wing slot slot when right clicked, before looking for other spaces in 'general slots'. Add this to the example?)

Solxanich committed:
Add SkipUIDraw to example mod instead of commenting out.

Solxanich committed:
Add Custom ApplyEquips support

direwolf420 commented:
If a mod would want to iterate over all accessories, including modded ones implemented through the PRs system, what code would it use?

Solxanich commented:
> If a mod would want to iterate over all accessories, including modded ones implemented through the PRs system, what code would it use?

Look at ItemSlot.tml.cs -> Accessory Swap. Gives a fairly clean iteration example

direwolf420 commented:
Is all the code in `AccessorySwap` method public so it can be used in other mods?
Also, the iteration cuts off at half the accessory length, what does that skip, vanity slots?

Chicken-Bones commented:
> If a mod would want to iterate over all accessories, including modded ones implemented through the PRs system, what code would it use?

I've long told people this is not the intended design, and actively avoided exposing the ability to do this, because modded slots normally have something 'special' about them. The vanilla (and tML) design is for mods to be able to trigger 'item effects' by calling the appropriate methods on `Player`, regardless of source, or whether the item actually exists.

direwolf420 commented:
There is a mod which requires full control over the accessory slots so it can block items going in/discard them if they somehow get in them anyway. That would require either hooking the tml methods using manual detours, or iterating over accessory slots.

https://github.com/hamstar0/tml-lockedabilities-mod/blob/57de8f32ef19c1f72579f9edc4eb3310cfa2c3c1/LockedAbilities/MyPlayer_Test_Armor.cs#L21

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/AccessorySlotLoader.cs:
```diff
@@ -0,0 +1,426 @@
+		/// DOES NOT affect vanilla behaviour of swapping items like for like where existing in a slot
+		/// </summary>
+		public void ModifyDefaultSwapSlot(Item item, ref int accSlotToSwapTo) {
+			for (int num = 0; num < ModSlotPlayer(Player).SlotCount(); num++) {
```
It might be nicer to run this iteration in reverse order? Otherwise, if I have 2 slots which are both 'priority' for an item, it will pick the last one. (Unless I do some weird check on `accSlotToSwapTo`)

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,116 @@
+
+		protected sealed override void Register() => Type = LoaderManager.Get<AccessorySlotLoader>().Register(this);
+
+		public bool IsEmpty() => FunctionalItem.IsAir && VanityItem.IsAir && DyeItem.IsAir;
```
might be nice for this, and `IsEnabled` to be properties for consistency?

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,116 @@
+		/// </summary>
+		public virtual void ApplyEquipEffects() {
+			Player.VanillaUpdateEquip(FunctionalItem);
+			Player.ApplyEquipFunctional(FunctionalItem, ShowVisuals);
```
`ApplyEquipFunctional` takes a second parameter `hideVisual`, so it should be `!ShowVisuals`

Solxanich replied:
Good catch! Property itself is mis-named, should be HideVisuals as true is invisible

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModAccessorySlot.cs:
```diff
@@ -0,0 +1,116 @@
+		/// DOES NOT affect vanilla behaviour of swapping items like for like where existing in a slot
+		/// Return true to set this slot as the default targetted slot.
+		/// </summary>
+		public virtual bool ModifyDefaultSwapSlot(Item item, int accSlotToSwapTo) => false;
```
I personally need a hook for when an item is quick swapped into my slot.

Solxanich replied:
An OnEnterSlot(context) hook for all movement vectors or just from quick swap?

Solxanich replied:
A hook seems unnecessary - why not just track on the AccessorySlot itself when FunctionalItem.Type changes?

Is it specifically that are looking for the context as to how it got in to the slot?

Solxanich replied:
A hook seems unnecessary - why not just track on the AccessorySlot itself when FunctionalItem.Type changes?

Is it specifically that are looking for the context as to how it got in to the slot?

Chicken-Bones commented on patches/tModLoader/Terraria/Main.cs.patch:
```diff
@@ -2740,6 +2761,73 @@
++
++				DrawDefenseCounter(randomLocationX, chkMax);
++
++				LoaderManager.Get<AccessorySlotLoader>().DrawAccSlots(num20);
```
Personal request. It would be great if there was a way to get access to `num20` for my slot draw positions. Maybe store it in a global variable in `Main`?

Solxanich replied:
You can (and should) use the existing xLoc and yLoc properties to specify where you want the slot to be drawn - from there, AccessorySlot.DrawModded() will provide the exact positions it recommends through its position argument.

That said, after saying that and doublechecking the code, I can see how having that info might be helpful in deciding where to draw said slot less blindly.

Added AccessorySlotLoader.DrawVerticalAlignment and corresponding summary

Chicken-Bones reviewed (changes requested):
Things I found while porting UtilitySlots! I'll probably have one more review/testing pass when you've implemented this feedback.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/AccessorySlotLoader.cs:
```diff
@@ -0,0 +1,426 @@
+		/// At the end, calls this.DrawModded() where you can override to have custom drawing code for visuals.
+		/// Also includes creating hidevisibilitybutton.
+		/// </summary>
+		internal void DrawFunctional(Item[] access, bool[] visbility, int context, int slot, bool flag3, int xLoc, int yLoc) {
```
Should use a `ref bool visibility` and drop the `slot` parameter

Solxanich replied:
DrawSlot() still needs a slot parameter, and propagates all the way to ItemSlot related functionality from this method.

So I've rewritten the method to just be drawing visibility so that the outcome makes sense.

Solxanich commented:
> Is all the code in `AccessorySwap` method public so it can be used in other mods?

Last I checked yes, should all be public.

> Also, the iteration cuts off at half the accessory length, what does that skip, vanity slots?
Yes, skips vanity slots, same pattern as armor.



> > If a mod would want to iterate over all accessories, including modded ones implemented through the PRs system, what code would it use?
> 
> I've long told people this is not the intended design, and actively avoided exposing the ability to do this, because modded slots normally have something 'special' about them. The vanilla (and tML) design is for mods to be able to trigger 'item effects' by calling the appropriate methods on `Player`, regardless of source, or whether the item actually exists.

I disagree, there are several scenarios where one might want to iterate over what players have equipped, such as in rpg-esque mods like what Dire pointed to. We should be encouraging using of a common data storage point, for benefit of mod compatibility in terms of what items the player has actively equipped/in use, independent of the item's actual effects. 
For example, let's say I want to do an accessory set bonus mod, similar to armor set bonuses. 
I would not care which slots have the exact items, just that said items are equipped as is someplace.

> There is a mod which requires full control over the accessory slots so it can block items going in/discard them if they somehow get in them anyway. That would require either hooking the tml methods using manual detours, or iterating over accessory slots.
> 
> https://github.com/hamstar0/tml-lockedabilities-mod/blob/57de8f32ef19c1f72579f9edc4eb3310cfa2c3c1/LockedAbilities/MyPlayer_Test_Armor.cs#L21

For what you're describing, ItemLoader.CanEquipAccessory() should be a sufficient hook.

Solxanich commented:
Generic review comment so I don't forget: 
So, to summarize, internalize capability to iterate, add an IncompatibleWith/IsEquivalentWith hook?

CanBeEquippedWith seems best?
on ItemLoader?

Solxanich commented on patches/tModLoader/Terraria/Player.TML.cs:
```diff
@@ -139,5 +140,61 @@ public partial class Player
+		/// Invoked at the end of loading vanilla player data from files to fix stuff that isn't initialized coming out of load.
+		/// Corrects the player.lavaMax time, wingsLogic, and no fall dmg to be accurate for the provided items in accessory slots.
+		/// </summary>
+		internal static void LoadPlayerLastMinuteFixes(Item item, Player newPlayer) {
```
Not a vanilla bug per se - the method really only runs during player select screen, so it doesn't really have any ability to know the slot is or isn't available.
I left it as private primarily because this method is more decorative than practical - it doesn't actually change anything in terms of rendering on player select nor functionality once in world (beyond the first tick).

I'll change it to public with the additional notes, but don't think we need to think too much about this method beyond that it has been matched for purposes of vanilla consistency.

Solxanich committed:
Resolve a few review comments

public-izing
Drop items on death
Better summaries
Renames
DrawCode design simplifications

Solxanich committed:
Add CanAccessoryBeEquippedWith() method

And add ItemLoader.CanEquipAccessory to several missing spots pertaining to left-click

Solxanich commented on patches/tModLoader/Terraria/Player.TML.cs:
```diff
@@ -139,5 +140,61 @@ public partial class Player
+
+			if (item.wingSlot > 0) {
+				if (invisible && (velocity.Y == 0f || mount.Active))
+					return;
```
|| This prevents the vanity accessory from rendering, unlike vanilla. The item in a wing's vanity slot may not be a wing.
Addressed.

|| Also, I think we could(should?) move this wingSlot check/hack to just check our new equipped wings field, and set wings to that, if the player has y velocity, before calling Player.UpdateVisibleAccessories.
Sounds reasonable, no point to have that in iterations when already storing the item itself

|| Both this method, and Player.UpdateVisibleAccessory should be public so mods can feed items into them.
Addressed

|| Really, this method should not take the vItem parameter. wings is set in UpdateVisibleAccessory anyway. && The iteration order for accessories is also important. All functional accessories should be iterated before all vanity accessories.
Why does the iteration order for accessories matter for functional before vanity? I personally can't see exactly what may break down at a glance, but is there something known that you are wary of?

Solxanich replied:
Nvm, see the issue. Will fix.

Solxanich committed:
Fix a Netcode bug & Verify #1832 is fixed on this branch

Solxanich committed:
Make Draw more robust wrt EquipPages

Solxanich committed:
Fix Wing Vanity issue in UpdateVisibleAccessories

Solxanich committed:
Merged Result - Player.cs false conflict

Solxanich committed:
Merge branch 'tModLoader-1.4' into 1.4_Extra_acc_fix_redux

Solxanich committed:
Add ModItem hooks

Solxanich committed:
Fix bad logical in last commit

Solxanich committed:
Logical Simplification in CanAccessoryBeEquippedWith

Chicken-Bones committed:
Fix ModAccessorySlotPlayer.Save/LoadData hooks after main branch merge

Solxanich committed:
Increase Robustness for multiple players

Chicken-Bones committed:
Provide ModAccessorySlot.Player context during ModPlayer loading

Chicken-Bones committed:
Partial support for mods hiding the default equip page.

Solxanich committed:
Merge pull request #60 from tModLoader/1.4

1.4

Solxanich committed:
Fix DefenseCounter and Drawing of scrollbar

Solxanich committed:
Merge branch '1.4_Extra_acc_fix_redux' of https://github.com/Solxanich/tModLoader into 1.4_Extra_acc_fix_redux

Solxanich committed:
Cache Itemslot.TML.cs Wing vanity/functional separation fix

Solxanich committed:
Fix Rendering order & Wings placement bug

Solxanich committed:
OnHover, HoverText Hooks; Vector2 position

Solxanich committed:
Switch CustomLocation to nullable Vector2

Chicken-Bones committed:
Fix merge conflict in patch

Solxanich committed:
Remove unnecessary ItemLoader Hook

Solxanich committed:
Revert "Fix merge conflict in patch"

This reverts commit 3722405e3d47493d457e399b5cf30758cf60e054.

Solxanich committed:
Replace Context with AccessorySlotType

Also fix slot drawing when one is skipped,
cleaning up MusicLoader code,
Post/Pre draw hooks,
And a partial ItemSlotContext Enum for record keeping sake

Chicken-Bones committed:
Add a more general implementation of static CurrentPlayer context
